lwt/client/dns_client_lwt.ml

```diff
@@ -103,38 +104,97 @@ module Transport : Dns_client.S
     Lwt_condition.wait t.timer_condition >>= fun () ->
     loop ()
 
+  let authenticator =
+    let _authenticator = ref None in
+    fun () ->
+      match !_authenticator with
+      | Some x -> x
+      | None -> match Ca_certs.authenticator () with
+        | Ok a -> _authenticator := Some a ; a
+        | Error `Msg m -> invalid_arg ("failed to load trust anchors: " ^ m)
+
+  let decode_resolv_conf data =
+    let ( let* ) = Result.bind in
+    let authenticator = authenticator () in
+    let* ns = Dns_resolvconf.parse data in
+    match
+      List.flatten
+        (List.map
+           (fun (`Nameserver ip) ->
+              let tls = Tls.Config.client ~authenticator ~ip () in
+              [ `Tls (tls, ip, 853) ; `Plaintext (ip, 53) ])
+           ns)
+    with
+    | [] -> Error (`Msg "no nameservers in resolv.conf")
+    | ns -> Ok ns
+
+  let resolv_conf () =
+    let ( let* ) = Result.bind in
+    let* data = read_file "/etc/resolv.conf" in
+    let* ns =
+      Result.map_error
+        (function `Msg msg ->
+           Log.warn (fun m -> m "error %s decoding resolv.conf %S" msg data);
+           `Msg msg)
+        (decode_resolv_conf data)
+    in
+    Ok (ns, Digest.string data)
+
+  let default_resolver () =
+    let authenticator = authenticator () in
+    let peer_name = Dns_client.default_resolver_hostname in
+    let tls_config = Tls.Config.client ~authenticator ~peer_name () in
+    List.flatten
+      (List.map (fun ip -> [
+             `Tls (tls_config, ip, 853); `Plaintext (ip, 53)
+           ]) Dns_client.default_resolvers)
+
+  let maybe_resolv_conf t =
+    let needs_update =
+      match read_file "/etc/resolv.conf", t.resolv_conf with
+      | Ok data, Some dgst ->
+        let dgst' = Digest.string data in
+        if Digest.equal dgst' dgst then
+          `No
+        else
+          `Data (data, dgst')
+      | Ok data, None ->
+        let digest = Digest.string data in
+        `Data (data, digest)
+      | Error _, None ->
+        `No
+      | Error `Msg msg, Some _ ->
+        Log.warn (fun m -> m "error reading /etc/resolv.conf: %s" msg);
+        `Default
+    in
+    match needs_update with
+    | `No -> ()
+    | `Default ->
+      t.resolv_conf <- None;
+      t.nameservers <- default_resolver ()
+    | `Data (data, dgst) ->
+      match decode_resolv_conf data with
+      | Ok ns ->
+        t.resolv_conf <- Some dgst;
+        t.nameservers <- ns
+      | Error `Msg msg ->
+        Log.warn (fun m -> m "error %s decoding resolv.conf: %S" msg data);
+        t.resolv_conf <- None;
+        t.nameservers <- default_resolver ()
+
   let create ?nameservers ~timeout () =
-    let nameservers =
+    let nameservers, resolv_conf =
       match nameservers with
       | Some (`Udp, _) -> invalid_arg "UDP is not supported"
-      | Some (`Tcp, ns) -> ns
+      | Some (`Tcp, ns) -> ns, None
```

Comment:
Perhaps we should preserve whether the nameservers were passed explicitly and then not try to read resolv.conf.